New execution module: textfsm

mirceaulinic opened the pull request:
### What does this PR do?

This is an execution module based on [Google TextFSM](https://github.com/google/textfsm), which extracts data from unstructured text.
Not among the best practices, but sometimes turns out that parsing raw text is the most you can do when working with archaic methodologies.

mirceaulinic committed:
New execution module: TextFSM

mirceaulinic committed:
Define index function and avoid redundant code

mirceaulinic committed:
More verbose debug logs

mirceaulinic committed:
Improve opts and provide doc

mirceaulinic committed:
Deps: documentation note and load condition

mirceaulinic committed:
Textfsm execution module autodoc

salt-jenkins commented:
@mirceaulinic, thanks for your PR! By analyzing the history of the files in this pull request, we identified @whiteinge, @jacobhammons and @rallytime to be potential reviewers.

Commit:
Fix obvious typo

mirceaulinic commented:
Thanks for helping out @cachedout - I was just about to push the change (and I've noticed few other small inconsistencies in the doc). :-)

mirceaulinic committed:
More doc typos and correct tests

mirceaulinic committed:
Lint

mirceaulinic committed:
Fix file handler bug

cachedout commented:
Go Go Jenkins!

cachedout commented on salt/modules/textfsm_mod.py:
```diff
@@ -0,0 +1,457 @@
+    HAS_CLITABLE = False
+
+# Import salt modules
+import salt.utils.files
```
You should be able to avoid this import and use the `__utils__` dictionary instead.

cachedout replied:
@mirceaulinic Did you see the above comment?

mirceaulinic replied:
Hi @cachedout yes, I saw your suggestion - thanks! I'll make this change too, it's in my long queue, but given that's a new module of Oxygen has a lower prio.
Don't worry, I will let you know when ready. Cheers!

mirceaulinic committed:
Import fopen function from salt.utils.files or salt.utils

The fopen function does not seem to be available in the
existing Salt releases, under the __utils__ dunder
(It will be in Oxygen as the `files.fopen` key).
For backwards compatibility and this module to be
able to be ported as extension modules when running
earlier releases than Oxygen, we can import the fopen
function in a try-except block, looking firstly to
import it from the new salt.utils.files module.

mirceaulinic commented:
Hi @cachedout - this PR should be ready now. I have diverged a little bit from your suggestion, but thanks to your idea I realised the import had to be improved anyway. Please see my comments in https://github.com/saltstack/salt/pull/42336/commits/d90ed050ba0744192cfb50a1e6d3292f6e5cc437.
Thanks!

rallytime commented:
re-run py3
